File: src/web/frontend/src/types/events.ts
/**
 * TypeScript discriminated union mirroring Python's SessionEvent models exactly.
 * One interface per event type — matches src/session/events.py.
 */

interface BaseEvent {
  timestamp: string; // ISO 8601
  turn_number: number;
  session_id: string;
}

export interface MessageEvent extends BaseEvent {
  type: 'MESSAGE';
  agent_id: string;
  agent_name: string;
  model: string;
  channel_id: string;
  recipient_id?: string;
  text: string;
  is_parallel: boolean;
}

export interface MonologueEvent extends BaseEvent {
  type: 'MONOLOGUE';
  agent_id: string;
  agent_name: string;
  text: string;
}

export interface TurnEvent extends BaseEvent {
  type: 'TURN';
  agent_ids: string[];
  is_parallel: boolean;
}

export interface GameStateEvent extends BaseEvent {
  type: 'GAME_STATE';
  updates: Record<string, unknown>;
  full_state: Record<string, unknown>;
}

export interface RuleViolationEvent extends BaseEvent {
  type: 'RULE_VIOLATION';
  agent_id: string;
  rule: string;
  violation_text: string;
}

export interface ChannelCreatedEvent {
  type: 'CHANNEL_CREATED';
  timestamp: string;
  session_id: string;
  channel_id: string;
  channel_type: 'public' | 'team' | 'private';
  members: string[];
}

export interface SessionEndEvent extends BaseEvent {
  type: 'SESSION_END';
  reason: 'max_turns' | 'win_condition' | 'completion_signal' | 'user_ended' | 'error';
  message?: string;
}

export interface IncidentEvent extends BaseEvent {
  type: 'INCIDENT';
  agent_id: string;
  agent_name: string;
  model: string;
  incident_type: 'timeout' | 'error';
  detail: string;
}

export type SessionEvent =
  | MessageEvent
  | MonologueEvent
  | TurnEvent
  | GameStateEvent
  | RuleViolationEvent
  | ChannelCreatedEvent
  | SessionEndEvent
  | IncidentEvent;

File: src/session/events.py
"""
Smart session event models.

Each event type is a distinct Pydantic model. SessionEvent is an
Annotated discriminated union — consumers get fully-typed access
to every field with no data: dict lookups.

Usage:
    from src.session.events import SessionEvent, MessageEvent, TurnEvent

    # Deserialize from raw dict (e.g., JSON sidecar, SSE payload)
    from pydantic import TypeAdapter
    ta = TypeAdapter(SessionEvent)
    event = ta.validate_python(raw_dict)

    # Type-narrowing in handlers
    match event.type:
        case "MESSAGE":
            print(event.channel_id, event.text)  # fully typed
        case "TURN":
            print(event.agent_ids, event.is_parallel)
"""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any, Literal

from pydantic import BaseModel, Field


class MessageEvent(BaseModel):
    """An agent (or HITL human) sent a message on a channel."""

    type: Literal["MESSAGE"] = "MESSAGE"
    timestamp: datetime
    turn_number: int
    session_id: str
    agent_id: str
    agent_name: str
    model: str
    channel_id: str  # "public", "team_red", "private_nova_rex", etc.
    recipient_id: str | None = None  # set only for private messages
    text: str
    is_parallel: bool = False  # True when generated during a parallel turn


class MonologueEvent(BaseModel):
    """An agent's internal chain-of-thought.

    NEVER injected into any other agent's context. Observer-only.
    Rendered in MonologuePanel (TUI) / Monologue Drawer (GUI).
    """

    type: Literal["MONOLOGUE"] = "MONOLOGUE"
    timestamp: datetime
    turn_number: int
    session_id: str
    agent_id: str
    agent_name: str
    text: str


class TurnEvent(BaseModel):
    """Signals the start of a new turn.

    Consumers (MonologuePanel, turn indicator) should reset/clear
    their state when this event fires.
    """

    type: Literal["TURN"] = "TURN"
    timestamp: datetime
    turn_number: int
    session_id: str
    agent_ids: list[str]  # agents about to speak this turn
    is_parallel: bool = False


class GameStateEvent(BaseModel):
    """Game state mutation after a turn (scores, round, win/loss, etc.)."""

    type: Literal["GAME_STATE"] = "GAME_STATE"
    timestamp: datetime
    turn_number: int
    session_id: str
    updates: dict[str, Any]  # key-value mutations applied this turn
    full_state: dict[str, Any]  # complete game state snapshot


class RuleViolationEvent(BaseModel):
    """An agent's response violated a game rule."""

    type: Literal["RULE_VIOLATION"] = "RULE_VIOLATION"
    timestamp: datetime
    turn_number: int
    session_id: str
    agent_id: str
    rule: str  # human-readable rule description
    violation_text: str  # excerpt of the violating response


class ChannelCreatedEvent(BaseModel):
    """A message channel was initialized at session start."""

    type: Literal["CHANNEL_CREATED"] = "CHANNEL_CREATED"
    timestamp: datetime
    session_id: str
    channel_id: str
    channel_type: Literal["public", "team", "private"]
    members: list[str]  # agent IDs; empty list means all agents


class SessionEndEvent(BaseModel):
    """The session has concluded."""

    type: Literal["SESSION_END"] = "SESSION_END"
    timestamp: datetime
    turn_number: int
    session_id: str
    reason: Literal[
        "max_turns",
        "win_condition",
        "completion_signal",
        "user_ended",
        "error",
    ]
    message: str | None = None  # optional human-readable summary


# ---------------------------------------------------------------------------
# Discriminated union — the single type used everywhere in the codebase
# ---------------------------------------------------------------------------

SessionEvent = Annotated[
    MessageEvent
    | MonologueEvent
    | TurnEvent
    | GameStateEvent
    | RuleViolationEvent
    | ChannelCreatedEvent
    | SessionEndEvent,
    Field(discriminator="type"),
]
